File: backend/MadrasaAccounting.Application/Services/IncomeService.cs
using MadrasaAccounting.Application.DTOs.Income;
using MadrasaAccounting.Domain.Entities;
using MadrasaAccounting.Infrastructure.Data;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using Newtonsoft.Json;

namespace MadrasaAccounting.Application.Services;

public class IncomeService : IIncomeService
{
    private readonly MadrasaDbContext _context;
    private readonly ILogger<IncomeService> _logger;

    public IncomeService(MadrasaDbContext context, ILogger<IncomeService> logger)
    {
        _context = context;
        _logger = logger;
    }

    public async Task<IncomeResponse> CreateIncomeAsync(Guid adminId, CreateIncomeRequest request)
    {
        using var transaction = await _context.Database.BeginTransactionAsync();

        try
        {
            // Validate category exists
            var category = await _context.IncomeCategories
                .FirstOrDefaultAsync(c => c.Id == request.CategoryId && c.AdminId == adminId);

            if (category == null)
                throw new InvalidOperationException("Category not found");

            // Create income record
            var income = new Income
            {
                AdminId = adminId,
                CategoryId = request.CategoryId,
                Amount = request.Amount,
                Date = request.Date,
                Description = request.Description,
                PaymentMethod = request.PaymentMethod,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow
            };

            _context.Incomes.Add(income);
            await _context.SaveChangesAsync();

            // Create audit log
            var auditLog = new AuditLog
            {
                AdminId = adminId,
                ActionType = "IncomeCreated",
                Description = $"Income created: {category.Name} - {request.Amount}",
                TransactionId = income.Id,
                NewValues = JsonConvert.SerializeObject(income),
                Timestamp = DateTime.UtcNow
            };

            _context.AuditLogs.Add(auditLog);
            await _context.SaveChangesAsync();

            await transaction.CommitAsync();

            _logger.LogInformation("Income created successfully. AdminId: {AdminId}, IncomeId: {IncomeId}", adminId, income.Id);

            return MapToResponse(income, category.Name);
        }
        catch (Exception ex)
        {
            await transaction.RollbackAsync();
            _logger.LogError(ex, "Error creating income for admin: {AdminId}", adminId);
            throw new Exception("An error occurred while creating income", ex);
        }
    }

    public async Task<IncomeResponse> GetIncomeByIdAsync(Guid adminId, Guid incomeId)
    {
        try
        {
            var income = await _context.Incomes
                .Include(i => i.Category)
                .FirstOrDefaultAsync(i => i.Id == incomeId && i.AdminId == adminId);

            if (income == null)
                throw new InvalidOperationException("Income not found");

            return MapToResponse(income, income.Category?.Name ?? "Unknown");
        }
        catch (InvalidOperationException)
        {
            throw;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error retrieving income: {IncomeId}", incomeId);
            throw new Exception("An error occurred while retrieving income", ex);
        }
    }

    public async Task<IncomePaginatedResponse> GetIncomeListAsync(Guid adminId, int pageNumber = 1, int pageSize = 10, DateTime? startDate = null, DateTime? endDate = null)
    {
        try
        {
            IQueryable<Income> query = _context.Incomes
                .Where(i => i.AdminId == adminId)
                .Include(i => i.Category);

            if (startDate.HasValue)
                query = query.Where(i => i.Date >= startDate.Value);

            if (endDate.HasValue)
                query = query.Where(i => i.Date <= endDate.Value);

            var totalCount = await query.CountAsync();

            var items = await query
                .OrderByDescending(i => i.Date)
                .Skip((pageNumber - 1) * pageSize)
                .Take(pageSize)
                .ToListAsync();

            return new IncomePaginatedResponse
            {
                Items = items.Select(i => MapToResponse(i, i.Category?.Name ?? "Unknown")).ToList(),
                TotalCount = totalCount,
                PageNumber = pageNumber,
                PageSize = pageSize
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error retrieving income list for admin: {AdminId}", adminId);
            throw new Exception("An error occurred while retrieving income list", ex);
        }
    }

    public async Task<IncomePaginatedResponse> SearchIncomeAsync(Guid adminId, string searchTerm, int pageNumber = 1, int pageSize = 10)
    {
        try
        {
            var query = _context.Incomes
                .Where(i => i.AdminId == adminId)
                .Include(i => i.Category)
                .Where(i => i.Description.Contains(searchTerm) || i.Category!.Name.Contains(searchTerm));

            var totalCount = await query.CountAsync();

            var items = await query
                .OrderByDescending(i => i.Date)
                .Skip((pageNumber - 1) * pageSize)
                .Take(pageSize)
                .ToListAsync();

            return new IncomePaginatedResponse
            {
                Items = items.Select(i => MapToResponse(i, i.Category?.Name ?? "Unknown")).ToList(),
                TotalCount = totalCount,
                PageNumber = pageNumber,
                PageSize = pageSize
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error searching income for admin: {AdminId}", adminId);
            throw new Exception("An error occurred while searching income", ex);
        }
    }

    public async Task<IncomeResponse> UpdateIncomeAsync(Guid adminId, Guid incomeId, UpdateIncomeRequest request)
    {
        using var transaction = await _context.Database.BeginTransactionAsync();

        try
        {
            var income = await _context.Incomes
                .Include(i => i.Category)
                .FirstOrDefaultAsync(i => i.Id == incomeId && i.AdminId == adminId);

            if (income == null)
                throw new InvalidOperationException("Income not found");

            // Validate category exists
            var category = await _context.IncomeCategories
                .FirstOrDefaultAsync(c => c.Id == request.CategoryId && c.AdminId == adminId);

            if (category == null)
                throw new InvalidOperationException("Category not found");

            // Store old values for audit
            var oldValues = JsonConvert.SerializeObject(new { income.Amount, income.Date, income.Description, income.PaymentMethod });

            // Update income
            income.CategoryId = request.CategoryId;
            income.Amount = request.Amount;
            income.Date = request.Date;
            income.Description = request.Description;
            income.PaymentMethod = request.PaymentMethod;
            income.UpdatedAt = DateTime.UtcNow;

            _context.Incomes.Update(income);
            await _context.SaveChangesAsync();

            // Create audit log
            var auditLog = new AuditLog
            {
                AdminId = adminId,
                ActionType = "IncomeUpdated",
                Description = $"Income updated: {category.Name} - {request.Amount}",
                TransactionId = incomeId,
                OldValues = oldValues,
                NewValues = JsonConvert.SerializeObject(new { income.Amount, income.Date, income.Description, income.PaymentMethod }),
                Timestamp = DateTime.UtcNow
            };

            _context.AuditLogs.Add(auditLog);
            await _context.SaveChangesAsync();

            await transaction.CommitAsync();

            _logger.LogInformation("Income updated successfully. AdminId: {AdminId}, IncomeId: {IncomeId}", adminId, incomeId);

            return MapToResponse(income, category.Name);
        }
        catch (InvalidOperationException)
        {
            throw;
        }
        catch (Exception ex)
        {
            await transaction.RollbackAsync();
            _logger.LogError(ex, "Error updating income: {IncomeId}", incomeId);
            throw new Exception("An error occurred while updating income", ex);
        }
    }

    public async Task DeleteIncomeAsync(Guid adminId, Guid incomeId)
    {
        using var transaction = await _context.Database.BeginTransactionAsync();

        try
        {
            var income = await _context.Incomes
                .Include(i => i.Category)
                .FirstOrDefaultAsync(i => i.Id == incomeId && i.AdminId == adminId);

            if (income == null)
                throw new InvalidOperationException("Income not found");

            var categoryName = income.Category?.Name ?? "Unknown";

            // Delete income
            _context.Incomes.Remove(income);
            await _context.SaveChangesAsync();

            // Create audit log
            var auditLog = new AuditLog
            {
                AdminId = adminId,
                ActionType = "IncomeDeleted",
                Description = $"Income deleted: {categoryName} - {income.Amount}",
                TransactionId = incomeId,
                OldValues = JsonConvert.SerializeObject(income),
                Timestamp = DateTime.UtcNow
            };

            _context.AuditLogs.Add(auditLog);
            await _context.SaveChangesAsync();

            await transaction.CommitAsync();

            _logger.LogInformation("Income deleted successfully. AdminId: {AdminId}, IncomeId: {IncomeId}", adminId, incomeId);
        }
        catch (InvalidOperationException)
        {
            throw;
        }
        catch (Exception ex)
        {
            await transaction.RollbackAsync();
            _logger.LogError(ex, "Error deleting income: {IncomeId}", incomeId);
            throw new Exception("An error occurred while deleting income", ex);
        }
    }

    private IncomeResponse MapToResponse(Income income, string categoryName)
    {
        return new IncomeResponse
        {
            Id = income.Id,
            CategoryId = income.CategoryId,
            CategoryName = categoryName,
            Amount = income.Amount,
            Date = income.Date,
            Description = income.Description ?? string.Empty,
            PaymentMethod = income.PaymentMethod,
            CreatedAt = income.CreatedAt,
            UpdatedAt = income.UpdatedAt
        };
    }
}

File: backend/src/services/incomeService.ts
import { Decimal } from '@prisma/client/runtime/library';
import prisma from '../config/database';
import { ApiError } from '../utils/errorHandler';
import logger from '../utils/logger';
import {
  CreateIncomeRequest,
  UpdateIncomeRequest,
  IncomeResponse,
  IncomeFilterParams,
} from '../types/income';

class IncomeService {
  /**
   * Create new income transaction
   */
  async createIncome(input: CreateIncomeRequest, userId: string): Promise<IncomeResponse> {
    try {
      // Validate input
      if (!input.date || !input.categoryId || input.amount <= 0) {
        throw new ApiError(400, 'Date, category, and amount (greater than zero) are required');
      }

      // Verify category exists
      const category = await prisma.incomeCategory.findUnique({
        where: { id: input.categoryId },
      });

      if (!category) {
        throw new ApiError(404, 'Income category not found');
      }

      // Create transaction
      const transaction = await prisma.transaction.create({
        data: {
          type: 'INCOME',
          date: new Date(input.date),
          incomeCategoryId: input.categoryId,
          description: input.description,
          amount: new Decimal(input.amount),
          paymentMethod: input.paymentMethod,
          reference: input.reference,
        },
        include: {
          incomeCategory: {
            select: { id: true, name: true },
          },
        },
      });

      // Log audit event
      await prisma.auditLog.create({
        data: {
          action: 'INCOME_CREATED',
          description: `Income created: ${input.description} - ${input.amount}`,
          userId,
          tableName: 'Transaction',
          recordId: transaction.id,
          newValues: JSON.parse(JSON.stringify(transaction)),
        },
      });

      logger.info(`Income created: ${transaction.id}`);

      return this.formatIncomeResponse(transaction);
    } catch (error) {
      if (error instanceof ApiError) {
        throw error;
      }
      logger.error('Error creating income', error as Error);
      throw new ApiError(500, 'Failed to create income');
    }
  }

  /**
   * Get income by ID
   */
  async getIncomeById(id: string): Promise<IncomeResponse> {
    try {
      const transaction = await prisma.transaction.findUnique({
        where: { id },
        include: {
          incomeCategory: {
            select: { id: true, name: true },
          },
        },
      });

      if (!transaction || transaction.type !== 'INCOME') {
        throw new ApiError(404, 'Income not found');
      }

      return this.formatIncomeResponse(transaction);
    } catch (error) {
      if (error instanceof ApiError) {
        throw error;
      }
      logger.error('Error fetching income', error as Error);
      throw new ApiError(500, 'Failed to fetch income');
    }
  }

  /**
   * Get all income with filters and pagination
   */
  async getIncomes(filters: IncomeFilterParams) {
    try {
      const {
        page = 1,
        pageSize = 10,
        startDate,
        endDate,
        categoryId,
        search,
        sortBy = 'date',
        sortOrder = 'desc',
      } = filters;

      // Build where clause
      const where: any = {
        type: 'INCOME',
      };

      if (startDate && endDate) {
        where.date = {
          gte: new Date(startDate),
          lte: new Date(endDate),
        };
      }

      if (categoryId) {
        where.incomeCategoryId = categoryId;
      }

      if (search) {
        where.OR = [
          { description: { contains: search, mode: 'insensitive' } },
          { reference: { contains: search, mode: 'insensitive' } },
        ];
      }

      // Get total count
      const total = await prisma.transaction.count({ where });

      // Get paginated results
      const items = await prisma.transaction.findMany({
        where,
        include: {
          incomeCategory: {
            select: { id: true, name: true },
          },
        },
        orderBy: {
          [sortBy]: sortOrder,
        },
        skip: (page - 1) * pageSize,
        take: pageSize,
      });

      const pages = Math.ceil(total / pageSize);

      return {
        items: items.map((item: any) => this.formatIncomeResponse(item)),
        total,
        page,
        pageSize,
        pages,
      };
    } catch (error) {
      logger.error('Error fetching incomes', error as Error);
      throw new ApiError(500, 'Failed to fetch incomes');
    }
  }

  /**
   * Update income transaction
   */
  async updateIncome(
    id: string,
    input: UpdateIncomeRequest,
    userId: string
  ): Promise<IncomeResponse> {
    try {
      // Get existing transaction
      const existing = await prisma.transaction.findUnique({
        where: { id },
      });

      if (!existing || existing.type !== 'INCOME') {
        throw new ApiError(404, 'Income not found');
      }

      // Verify category if provided
      if (input.categoryId) {
        const category = await prisma.incomeCategory.findUnique({
          where: { id: input.categoryId },
        });

        if (!category) {
          throw new ApiError(404, 'Income category not found');
        }
      }

      // Update transaction
      const updated = await prisma.transaction.update({
        where: { id },
        data: {
          ...(input.date && { date: new Date(input.date) }),
          ...(input.categoryId && { incomeCategoryId: input.categoryId }),
          ...(input.description && { description: input.description }),
          ...(input.amount && { amount: new Decimal(input.amount) }),
          ...(input.paymentMethod && { paymentMethod: input.paymentMethod }),
          ...(input.reference !== undefined && { reference: input.reference }),
        },
        include: {
          incomeCategory: {
            select: { id: true, name: true },
          },
        },
      });

      // Log audit event
      await prisma.auditLog.create({
        data: {
          action: 'INCOME_UPDATED',
          description: `Income updated: ${updated.description}`,
          userId,
          tableName: 'Transaction',
          recordId: id,
          oldValues: JSON.parse(JSON.stringify(existing)),
          newValues: JSON.parse(JSON.stringify(updated)),
        },
      });

      logger.info(`Income updated: ${id}`);

      return this.formatIncomeResponse(updated);
    } catch (error) {
      if (error instanceof ApiError) {
        throw error;
      }
      logger.error('Error updating income', error as Error);
      throw new ApiError(500, 'Failed to update income');
    }
  }

  /**
   * Delete income transaction
   */
  async deleteIncome(id: string, userId: string): Promise<void> {
    try {
      // Get existing transaction
      const existing = await prisma.transaction.findUnique({
        where: { id },
      });

      if (!existing || existing.type !== 'INCOME') {
        throw new ApiError(404, 'Income not found');
      }

      // Delete transaction
      await prisma.transaction.delete({
        where: { id },
      });

      // Log audit event
      await prisma.auditLog.create({
        data: {
          action: 'INCOME_DELETED',
          description: `Income deleted: ${existing.description}`,
          userId,
          tableName: 'Transaction',
          recordId: id,
          oldValues: JSON.parse(JSON.stringify(existing)),
        },
      });

      logger.info(`Income deleted: ${id}`);
    } catch (error) {
      if (error instanceof ApiError) {
        throw error;
      }
      logger.error('Error deleting income', error as Error);
      throw new ApiError(500, 'Failed to delete income');
    }
  }

  /**
   * Get income summary for a date range
   */
  async getIncomeSummary(startDate?: string, endDate?: string) {
    try {
      const where: any = {
        type: 'INCOME',
      };

      if (startDate && endDate) {
        where.date = {
          gte: new Date(startDate),
          lte: new Date(endDate),
        };
      }

      const summary = await prisma.transaction.aggregate({
        where,
        _sum: {
          amount: true,
        },
        _count: {
          id: true,
        },
      });

      return {
        totalIncome: summary._sum.amount || 0,
        count: summary._count.id || 0,
      };
    } catch (error) {
      logger.error('Error getting income summary', error as Error);
      throw new ApiError(500, 'Failed to get income summary');
    }
  }

  /**
   * Format transaction response
   */
  private formatIncomeResponse(transaction: any): IncomeResponse {
    return {
      id: transaction.id,
      date: transaction.date.toISOString().split('T')[0],
      categoryId: transaction.incomeCategoryId || '',
      category: transaction.incomeCategory
        ? {
            id: transaction.incomeCategory.id,
            name: transaction.incomeCategory.name,
          }
        : undefined,
      description: transaction.description,
      amount: parseFloat(transaction.amount.toString()),
      paymentMethod: transaction.paymentMethod,
      reference: transaction.reference,
      createdAt: transaction.createdAt.toISOString(),
      updatedAt: transaction.updatedAt.toISOString(),
    };
  }
}

export default new IncomeService();
